pinot-core/src/main/java/org/apache/pinot/core/data/table/SimpleIndexedTable.java

```diff
@@ -139,6 +144,14 @@ public int size() {
 
   @Override
   public Iterator<Record> iterator() {
+    if (_iterator == null) {
```

Comment:
The existing implementation seems to be written assuming that clients won't have to iterate on the table until they call finish() and that is why the iterator() simply returned the iterator (NULL if client didn't call finish prior to it). So basically the workflow imposed on the client is:

-- upsert() as many as you want
-- finish()
-- iterator()
-- done 

and finish() should only be called once as per the Javadoc of this API.

Now you may still want to simply iterate (over random records) without calling finish() -- finish() will have a cost of resizing for ORDER BY queries. We have this need during merge of DistinctAggregationFunction() where we iterate over 1 ... N-1 indexed tables and merge them into 0th indexed table.  Only after the merge is done, we call finish() on the 0th indexed table since that has all the records. That way we get the final iterator() and trimmed set.

Now in iterator() function, we can't just simply do _lookupMap.values().iterator() since if the user called finish() prior to it and there was order by, then finish() would have created an iterator over sorted records. So now doing _lookupMap.values().iterator() will give incorrect order to the caller. 

I agree with the fact that user of a data structure should be able to create an iterator as many times as they want. However, for IndexedTable the creation of iterator is tied to finish().

Reply:
if you do this, you allow the  possibility that the _lookupMap gets modified while you are iterating

Reply:
To elaborate some more, one caller could get an iterator and be iterating, another caller could be upserting/resizing, and we get ConcurrentModificationEx.
Also we seem to br breaking the contract of upsert-> finish->iterator, but only for SimpleTable.
As per our offline conversation, how about instead we still keep that contract, but add the control to finish(), to skip the resizing. finish() already has the controls to skip ordering, this can be another one

Reply:
ConcurrentModificationException will not happen since Distinct code controlls the usage of SimpleIndexedTable.

Reply:
Yes we can either augment finish() by passing additional argument to tell it not do anything except create an iterator or for now we can call finish() first before the server combine and evaluate later. I am more inclined towards changing the API.